linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala

```diff
@@ -0,0 +1,964 @@
+      reportError("Top level export method cannot have the flag Mutable")
+    if (flags.namespace != MemberNamespace.PublicStatic)
+      reportError("Top level export must be public and static")
+
+    if (!pName.isInstanceOf[StringLiteral])
+      reportError("Top level exports may not have computed names")
+
+    checkJSParamDefs(params, restParam)
+    checkTree(body, Env.fromParams(params ++ restParam))
+  }
+
+  private def checkTopLevelFieldExportDef(
+      topLevelFieldExportDef: TopLevelFieldExportDef): Unit = {
+    implicit val ctx = ErrorContext(topLevelFieldExportDef)
+
+    if (!classDef.kind.isAnyNonNativeClass)
+      reportError("native classes may not have field exports")
+
+    val field = topLevelFieldExportDef.field
+
+    fields(MemberNamespace.PublicStatic.ordinal).get(field.name).fold {
+      reportError(i"Cannot export non-existent static field '$field'")
+    } { tpe =>
+      if (tpe != AnyType)
+        reportError(i"Cannot export field '$field' of type $tpe")
+    }
+  }
+
+  private def checkMethodNameNamespace(name: MethodName, namespace: MemberNamespace)(
+      implicit ctx: ErrorContext): Unit = {
+    if (name.isReflectiveProxy && namespace != MemberNamespace.Public)
+      reportError("reflective proxies must be in the public (non-static) namespace")
+
+    if (name.isConstructor != (namespace == MemberNamespace.Constructor))
+      reportError("a member can have a constructor name iff it is in the constructor namespace")
+
+    if ((name.isStaticInitializer || name.isClassInitializer) != (namespace == MemberNamespace.StaticConstructor))
+      reportError("a member can have a static constructor name iff it is in the static constructor namespace")
+  }
+
+  private def checkJSParamDefs(params: List[ParamDef], restParam: Option[ParamDef])(
+      implicit ctx: ErrorContext): Unit = {
+    for (ParamDef(name, _, ptpe, _) <- params ++ restParam) {
+      checkDeclareLocalVar(name)
+      if (ptpe != AnyType)
+        reportError(i"Parameter $name has type $ptpe but must be any")
+    }
+  }
+
+  private def checkTreeOrSpreads(trees: List[TreeOrJSSpread], env: Env): Unit = {
+    trees.foreach {
+      case JSSpread(items) => checkTree(items, env)
+      case tree: Tree      => checkTree(tree, env)
+    }
+  }
+
+  private def checkTrees(trees: List[Tree], env: Env): Unit =
+    trees.foreach(checkTree(_, env))
+
+  private def checkTree(tree: Tree, env: Env): Env = {
+    implicit val ctx = ErrorContext(tree)
+
+    def checkApplyGeneric(methodName: MethodName, args: List[Tree]): Unit = {
+      val paramRefs = methodName.paramTypeRefs.size
+      if (args.size != paramRefs)
+        reportError(i"Arity mismatch: $paramRefs expected but "+
+            i"${args.size} found")
+      checkTrees(args, env)
+    }
+
+    tree match {
+      case VarDef(ident, _, vtpe, mutable, rhs) =>
+        checkDeclareLocalVar(ident)
+        checkTree(rhs, env)
+        env.withLocal(LocalDef(ident.name, vtpe, mutable))
```

Comment:
This is the only `case` that returns an env different from `env`. Could we put that separately from the main `tree match`?

Reply:
Done.